### PR TITLE
Doc onAppBooting

### PR DESCRIPTION
**Pull request by akanex115**
# Background
#40

## TODOs
- [x] Write document for src/hooks/app/use-is-app-booted.hook.ts

## Related PRs
- _None_

## Checklist (Developers)
- [x] Title is checked
- [x] Background & TODOs are filled
- [x] `Related PRs` is correctly modified, if it is not `None`
- [x] Assignee is set
- [x] Labels are set
- [x] Appropriate issue(s) is(are) linked to this PR under `development`
- [x] `yarn inspect` is run
- [x] `TODOs` are handled and checked
- [x] Final Operation Check is done

## Checklist (Reviewers)
- [x] Check if there are any other missing TODOs that are not yet listed
- [x] Review Code
- [x] Every item on the checklist has been addressed accordingly
- [x] If `development` is associated to this PR, you must check if every TODOs are handled

### CODE REVIEW

File: dev/diagrams/on-app-booting-hook.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+activate user
+	user -> fe: open app
+	activate fe
+		fe -> hook: useIsAppBooted
+		activate hook
+			hook -> authPrep: onGetAuthPrep
+			activate authPrep
+				break if there is no auth prep or the user is not signed in
+					hook <- authPrep: throw new Error('Not Singed In')
```
Comment (mlajkim) [inline dev/diagrams/on-app-booting-hook.md:40]:
```suggestion
					hook <- authPrep: throw new Error('Not Signed In')
```

File: dev/diagrams/on-app-booting-hook.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+					fe <- hook: router.push(DEFAULT_MAIN_APP_PAGE)
+				end break
+				hook <- authPrep: handleSignOutApp()
+			deactivate authPrep
+			hook -> recoil: useEffect()
+			activate recoil
+				break if isBooted is true
+			  	hook <- recoil: Returns onAppBooting()
+				end break
+			deactivate recoil
```
Comment (mlajkim) [inline dev/diagrams/on-app-booting-hook.md:50]:
```suggestion
```
Can you explain what `useEffect()` is by 1/23?
https://react.dev/reference/react/useEffect

File: dev/diagrams/on-app-booting-hook.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+box "FE" #Lightblue
+  participant "Component" as fe
+  participant "Hook: useIsAppBooted" as hook
+	participant "UseAuthPrepHook" as authPrep
+	participant "Recoil" as recoil
+box
+box "API" #Lightgreen
+  participant "API" as api
+box
```
Comment (mlajkim) [inline dev/diagrams/on-app-booting-hook.md:30]: Is API Used?

File: dev/diagrams/on-app-booting-hook.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+				break if isBooted is true
+			  	hook <- recoil: Returns onAppBooting()
+				end break
+			deactivate recoil
+			fe <- hook: Returns nothing
+		deactivate hook
+	  user <- fe: Returns nothing
+	deactivate fe
+deactivate user
+
```
Comment (mlajkim) [inline dev/diagrams/on-app-booting-hook.md:56]: How would you explain `await handleSignOutApp()`

**Review by mlajkim — COMMENTED** [review]
First review

**Comment by mlajkim** [comment]
Expected:
image

File: dev/diagrams/on-app-booting-hook.md (hunk 3 above)
Comment (akanex115) [reply dev/diagrams/on-app-booting-hook.md:29]: Handled in https://github.com/ajktown/docs/pull/45/commits/7ca9a41e33ad46b2022d108fef64b5726fff8901

File: dev/diagrams/on-app-booting-hook.md (hunk 2 above)
Comment (akanex115) [reply dev/diagrams/on-app-booting-hook.md]: useEffect() is used to achieve side effects within a component.
It is executed after the component has been drawn and can monitor changes in the state of certain values.

File: dev/diagrams/on-app-booting-hook.md (hunk 4 above)
Comment (akanex115) [reply dev/diagrams/on-app-booting-hook.md:57]: await handleSignOutApp()はuseOnSignOutAppHookを利用している。
サインインされていない場合にサインアウト処理を行う。

1. APIへサインアウトのリクエストを送信。リクエストが完了するまで待機。
2. reset(wordIdsState)、reset(preferenceState)の実行。WordIDやユーザーの設定に関するStateを初期化する。
3. router.push(PageConst.Welcome)の実行。Welcome ページにリダイレクト。

File: dev/diagrams/on-app-booting-hook.md (hunk 2 above)
Comment (mlajkim) [reply dev/diagrams/on-app-booting-hook.md]: nice.

File: dev/diagrams/on-app-booting-hook.md (hunk 3 above)
Comment (mlajkim) [reply dev/diagrams/on-app-booting-hook.md:29]: @akanex115 
image

今回主人公は`useIsAppBooted()`だから、API通信をするuseAuthPrepHookは省略でいいかも。(APIも無くしていいかも)

File: dev/diagrams/on-app-booting-hook.md (hunk 4 above)
Comment (mlajkim) [reply dev/diagrams/on-app-booting-hook.md:57]: ここで、「APIへサインアウトのリクエストを送信。リクエストが完了するまで待機。」だけど、
現状実装上物理的にSign outさせるのができるのはAPIサーバーしかない。FE一人では、できない。(Securityのため）

https://zenn.dev/lilac31/articles/c0f216d29ac2b9

Comment (mlajkim) [reply dev/diagrams/on-app-booting-hook.md:57]: Resolvable

File: dev/diagrams/on-app-booting-hook.md (hunk 3 above)
Comment (akanex115) [reply dev/diagrams/on-app-booting-hook.md]: Handled in [6dac4a9](https://github.com/ajktown/docs/pull/45/commits/6dac4a995ccb75a36e53fe4aa5e74f9e35c3b2b9)

File: dev/diagrams/on-app-booting-hook.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  participant "Hook: useIsAppBooted" as hook
+box
+
+activate user
+  user -> fe: open app
+  activate fe
+    fe -> hook: useIsAppBooted
+    activate hook
+      hook <- hook: handleSignOutApp()
+      hook <- hook: router.push(DEFAULT_MAIN_APP_PAGE)
```
Comment (mlajkim) [inline dev/diagrams/on-app-booting-hook.md:33]:
```suggestion
      hook -> hook: handleSignOutApp()
      hook -> hook: router.push(DEFAULT_MAIN_APP_PAGE)
```

Better to make it `a -> a`

Good Example:
image

Bad example:
image

File: dev/diagrams/on-app-booting-hook.md [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      fe <- hook: Returns nothing
+    deactivate hook
+    user <- fe: Returns nothing
+  deactivate fe
+deactivate user
+
+
+
+
+
```
Comment (mlajkim) [inline dev/diagrams/on-app-booting-hook.md:43]:
```suggestion
```
Only one indent at the end of the file!!

**Review by mlajkim — COMMENTED** [review]
First review done
TODOs
- [ ] When does the `handleSignOutApp()` happens? (write it with break)
- [ ] Understand `try, catch, finally`
  - https://www.javadrive.jp/start/exception/index3.html

image

**Comment by akanex115** [comment]
The timing of when the `handleSignOutApp()` occurs is described.
Handled in [71a8192](https://github.com/ajktown/docs/pull/45/commits/71a8192e9aed0e35097742c5de22d52b8296d02b)

File: dev/diagrams/on-app-booting-hook.md [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  participant "Hook: useIsAppBooted" as hook
+box
+
+activate user
+  user -> fe: open app
+  activate fe
+    fe -> hook: useIsAppBooted
+    activate hook
+      break if there is no authPrep or the user is not signed in
+        hook -> hook: handleSignOutApp()
```
Comment (mlajkim) [inline dev/diagrams/on-app-booting-hook.md:33]:
```suggestion
        hook -> hook: handleSignOutApp()
        fe <- hook: Returns nothing
```
The current flow still makes the `` happens as the red line.
image


You need to make sure that the break ends at the component level!
image

**Review by mlajkim — COMMENTED** [review]
Review done

**Review by mlajkim — APPROVED** [review]
OK

**Pull request merged**